### Beacttini et al, CHM
### codes for analyses and plots shown in Figures:
### Figure 1E, Figure 2C
### Figure 3A, Supplementary Figure 3A
### Figure 3B
### Figures 3C, Supplementary Figure 3B
### Figures 1D, 2B
### Supplementary Figures 1E, 2E


##### load needed libraries ####
library(RColorBrewer)
library(tibble)
library(pheatmap)
library(Rsubread)
library(clusterProfiler)
library(stringr)
library(ggplot2)
library(grDevices)
library(genefilter)
library(xlsx)
library(tidyr)
library(openxlsx)
library(reshape2)
library(ggrepel)
library(gplots)
library(ggfortify)
library(cluster)
library(cowplot)
library(ggpubr)
library(ggsci)
library(dplyr)
library(DESeq2)

setwd('/Users/becattis/Desktop/CHM revision/Scripts to be uploaded/CBBP-transcriptomics/Transcriptomics is SPF CBBP (aCD3 and flagellin)/')
#save.image('CBBP_main.RData')
load('CBBP_main.RData')
### to run this need .gff files in working directory, 
### as well as a 'SAM_BAM_ALIGNMENTS' folder that contains all bowtie2 alignments files (.bam) obtained with the enclosed bash script
### the first step (FeatureCounts) can be avoided by directly reading in the relative df, here provided as .txt tables

# prepare .gff files 

BP <- read.delim("BP.gff", header=T)
BS <- read.delim("BS.gff", header=T)
CB <- read.delim("CB.gff", header=T)
PD <- read.delim("PD.gff", header=T)


# run the following if aligning raw reads from scratch using the provided bash script, skip otherwise
#BS_list <- list.files("/MyFolder/SAM_BAM_ALIGNMENTS", pattern = "BS.only_aligned.bam")
#BP_list <- list.files("/MyFolder/SAM_BAM_ALIGNMENTS", pattern = "BP.only_aligned.bam")
#CB_list <- list.files("/MyFolder/SAM_BAM_ALIGNMENTS", pattern = "CB.only_aligned.bam")
#PD_list <- list.files("/MyFolder/SAM_BAM_ALIGNMENTS", pattern = "PD.only_aligned.bam")


#### the following codes generate raw data for each CBBP members 
#### following are codes to generate cumulative figures



##### BP Count ####

## count aligning reads using Feature counts (skip if simply reading in the relative .txt)

#BP_FtCounts <- featureCounts(BP_list, annot.ext = BP, isGTFAnnotationFile=FALSE, largestOverlap =T, countChimericFragments=TRUE)
#BP_Counts <- as.data.frame(BP_FtCounts$counts) %>% select(-1)

#colnames(BP_Counts) <- str_replace(colnames(BP_Counts), '_sb93', '')
#colnames(BP_Counts) <- str_extract(colnames(BP_Counts), regex(pattern="Sample_m[0-9]{1,2}_[0-9]{1,2}h_flag|Sample_m[0-9]{1,2}_[0-9]{1,2}h_aCD3|Sample_m[0-9]{1}_0h"))
#colnames(BP_Counts) <- str_replace(colnames(BP_Counts), 'Sample_', 'BP_') 

## reorder columns
#BP_Counts <- BP_Counts[,c("BP_m1_0h", "BP_m2_0h", "BP_m3_0h", "BP_m4_6h_flag", "BP_m5_6h_flag", "BP_m6_6h_flag", "BP_m7_6h_aCD3", "BP_m8_6h_aCD3", "BP_m9_6h_aCD3", "BP_m10_24h_flag", "BP_m11_24h_flag" , "BP_m12_24h_flag" , "BP_m13_24h_aCD3", "BP_m14_24h_aCD3", "BP_m15_24h_aCD3")]

## remove tRNA/rRNA reads
#BP_Counts$feature <- row.names(BP_Counts)
#BP_Counts <- BP_Counts %>% filter(!grepl(pattern="tRNA-[A-Z, a-z]{3}-[A-Z, a-z]{3}" , BP_Counts$feature))
#BP_Counts <- BP_Counts %>% filter(!grepl(pattern="LSU rRNA|SSU rRNA|5S rRNA" , BP_Counts$feature))
#rownames(BP_Counts) <- BP_Counts$feature 
#BP_Counts$feature <- NULL

## read in the counts table generated by the above script

BP_Counts <- read.delim2('SB093_BP_Counts.txt', header = T, sep='\t') 
BS_Counts <- read.delim2('SB093_BS_Counts.txt', header = T, sep='\t') 
CB_Counts <- read.delim2('SB093_CB_Counts.txt', header = T, sep='\t') 
PD_Counts <- read.delim2('SB093_PD_Counts.txt', header = T, sep='\t') 

##### Run DeSeq2 ####
## BP
BP_colData <- as.data.frame(matrix(c("0h", "0h", "0h", "6h_flag","6h_flag", "6h_flag", "6h_aCD3", "6h_aCD3", "6h_aCD3" ,"24h_flag", "24h_flag", "24h_flag","24h_aCD3", "24h_aCD3", "24h_aCD3"),nrow=15,ncol=1))
row.names(BP_colData) <-c("BP_m1_0h", "BP_m2_0h", "BP_m3_0h", "BP_m4_6h_flag", "BP_m5_6h_flag", "BP_m6_6h_flag", "BP_m7_6h_aCD3", "BP_m8_6h_aCD3", "BP_m9_6h_aCD3", "BP_m10_24h_flag", "BP_m11_24h_flag" , "BP_m12_24h_flag" , "BP_m13_24h_aCD3", "BP_m14_24h_aCD3", "BP_m15_24h_aCD3")
colnames(BP_colData)<-c('BP_condition')
BP_countData <-  as.matrix(select(BP_Counts, BP_m1_0h, BP_m2_0h, BP_m3_0h, BP_m4_6h_flag, BP_m5_6h_flag, BP_m6_6h_flag, BP_m7_6h_aCD3, BP_m8_6h_aCD3, BP_m9_6h_aCD3, BP_m10_24h_flag, BP_m11_24h_flag , BP_m12_24h_flag , BP_m13_24h_aCD3, BP_m14_24h_aCD3, BP_m15_24h_aCD3), rownames.force=T )
## run Deseq excluding low count entries
BP_dds <- DESeqDataSetFromMatrix(countData=BP_countData, colData=BP_colData, design= ~ BP_condition)
BP_dds <- BP_dds[rowSums(counts(BP_dds)) > 10, ]
BP_dds <- DESeq(BP_dds, fitType = 'parametric')

BP_table_counts_normalized <- counts(BP_dds, normalized=TRUE)

## BS
BS_colData <- as.data.frame(matrix(c("0h", "0h", "0h", "6h_flag","6h_flag", "6h_flag", "6h_aCD3", "6h_aCD3", "6h_aCD3" ,"24h_flag", "24h_flag", "24h_flag","24h_aCD3", "24h_aCD3", "24h_aCD3"),nrow=15,ncol=1))
row.names(BS_colData) <-c("BS_m1_0h", "BS_m2_0h", "BS_m3_0h", "BS_m4_6h_flag", "BS_m5_6h_flag", "BS_m6_6h_flag", "BS_m7_6h_aCD3", "BS_m8_6h_aCD3", "BS_m9_6h_aCD3", "BS_m10_24h_flag", "BS_m11_24h_flag" , "BS_m12_24h_flag" , "BS_m13_24h_aCD3", "BS_m14_24h_aCD3", "BS_m15_24h_aCD3")
colnames(BS_colData)<-c('BS_condition')
BS_countData <-  as.matrix(select(BS_Counts, BS_m1_0h, BS_m2_0h, BS_m3_0h, BS_m4_6h_flag, BS_m5_6h_flag, BS_m6_6h_flag, BS_m7_6h_aCD3, BS_m8_6h_aCD3, BS_m9_6h_aCD3, BS_m10_24h_flag, BS_m11_24h_flag , BS_m12_24h_flag , BS_m13_24h_aCD3, BS_m14_24h_aCD3, BS_m15_24h_aCD3), rownames.force=T )

## run Deseq excluding low count entries
BS_dds <- DESeqDataSetFromMatrix(countData=BS_countData, colData=BS_colData, design= ~ BS_condition)
BS_dds <- BS_dds[rowSums(counts(BS_dds)) > 10, ]
BS_dds <- DESeq(BS_dds, fitType = 'parametric')

BS_table_counts_normalized <- counts(BS_dds, normalized=TRUE)

## CB
CB_colData <- as.data.frame(matrix(c("0h", "0h", "0h", "6h_flag","6h_flag", "6h_flag", "6h_aCD3", "6h_aCD3", "6h_aCD3" ,"24h_flag", "24h_flag", "24h_flag","24h_aCD3", "24h_aCD3", "24h_aCD3"),nrow=15,ncol=1))
row.names(CB_colData) <-c("CB_m1_0h", "CB_m2_0h", "CB_m3_0h", "CB_m4_6h_flag", "CB_m5_6h_flag", "CB_m6_6h_flag", "CB_m7_6h_aCD3", "CB_m8_6h_aCD3", "CB_m9_6h_aCD3", "CB_m10_24h_flag", "CB_m11_24h_flag" , "CB_m12_24h_flag" , "CB_m13_24h_aCD3", "CB_m14_24h_aCD3", "CB_m15_24h_aCD3")
colnames(CB_colData)<-c('CB_condition')
CB_countData <-  as.matrix(select(CB_Counts, CB_m1_0h, CB_m2_0h, CB_m3_0h, CB_m4_6h_flag, CB_m5_6h_flag, CB_m6_6h_flag, CB_m7_6h_aCD3, CB_m8_6h_aCD3, CB_m9_6h_aCD3, CB_m10_24h_flag, CB_m11_24h_flag , CB_m12_24h_flag , CB_m13_24h_aCD3, CB_m14_24h_aCD3, CB_m15_24h_aCD3), rownames.force=T )
## run Deseq excluding low count entries
CB_dds <- DESeqDataSetFromMatrix(countData=CB_countData, colData=CB_colData, design= ~ CB_condition)
CB_dds <- CB_dds[rowSums(counts(CB_dds)) > 10, ]
CB_dds <- DESeq(CB_dds, fitType = 'parametric')

CB_table_counts_normalized <- counts(CB_dds, normalized=TRUE)

as.data.frame(CB_table_counts_normalized) %>% rownames_to_column('Gene') %>% write.table('CB_norm_counts_DDS.txt', sep='\t', quote=F)
## PD
PD_colData <- as.data.frame(matrix(c("0h", "0h", "0h", "6h_flag","6h_flag", "6h_flag", "6h_aCD3", "6h_aCD3", "6h_aCD3" ,"24h_flag", "24h_flag", "24h_flag","24h_aCD3", "24h_aCD3", "24h_aCD3"),nrow=15,ncol=1))
row.names(PD_colData) <-c("PD_m1_0h", "PD_m2_0h", "PD_m3_0h", "PD_m4_6h_flag", "PD_m5_6h_flag", "PD_m6_6h_flag", "PD_m7_6h_aCD3", "PD_m8_6h_aCD3", "PD_m9_6h_aCD3", "PD_m10_24h_flag", "PD_m11_24h_flag" , "PD_m12_24h_flag" , "PD_m13_24h_aCD3", "PD_m14_24h_aCD3", "PD_m15_24h_aCD3")
colnames(PD_colData)<-c('PD_condition')
PD_countData <-  as.matrix(select(PD_Counts, PD_m1_0h, PD_m2_0h, PD_m3_0h, PD_m4_6h_flag, PD_m5_6h_flag, PD_m6_6h_flag, PD_m7_6h_aCD3, PD_m8_6h_aCD3, PD_m9_6h_aCD3, PD_m10_24h_flag, PD_m11_24h_flag , PD_m12_24h_flag , PD_m13_24h_aCD3, PD_m14_24h_aCD3, PD_m15_24h_aCD3), rownames.force=T )
## run Deseq excluding low count entries
PD_dds <- DESeqDataSetFromMatrix(countData=PD_countData, colData=PD_colData, design= ~ PD_condition)
PD_dds <- PD_dds[rowSums(counts(PD_dds)) > 10, ]
PD_dds <- DESeq(PD_dds, fitType = 'parametric')

PD_table_counts_normalized <- counts(PD_dds, normalized=TRUE)


##### PCA Figure 3A and Supplementary Figure 3A ####

## BP
BP_rld <- rlog(BP_dds, blind=FALSE)
BP_norm_counts <- as.data.frame(assay(BP_rld))

# visualize PCA using all counts
Supplementary_FIgure_3A_BP <- plotPCA(BP_rld, intgroup=c("BP_condition")) + theme_bw(base_size = 18) + geom_point(size=4) + theme(aspect.ratio = 1) + ggtitle('BP') + geom_text_repel(aes(label = name)) 

## make a custom PCAwith only significantly different genes using autoplot

# select only genes with a significant difference in any comparison
BP_LRT_dds <- DESeq(BP_dds, test="LRT", reduced=~1) 
BP_res <- as.data.frame(results(BP_LRT_dds))
BP_res$gene <- rownames(BP_res)
# make a vector with names of those significant genes
BP_res_sign <- filter(BP_res, padj < .05 & abs(log2FoldChange) > 1) 
BP_res_sign <- BP_res_sign$gene 
BP_norm_counts_LRT <- BP_norm_counts[which(rownames(BP_norm_counts) %in% BP_res_sign), ] 
BP_LRT_prcomp <- prcomp(x=t(BP_norm_counts_LRT), center = T, scale=T) #run PCA on the selected table of counts 
t_BP_norm_counts_LRT <- as.data.frame(t(BP_norm_counts_LRT)) %>% 
  mutate(Time_point=str_extract(rownames(BP_LRT_prcomp$x), pattern="0h|6h|24h"))  %>% 
  mutate(Treatment=str_extract(rownames(BP_LRT_prcomp$x), pattern="aCD3|flag")) %>% 
  replace_na(list(Treatment="None")) 


Figure_3A_BP <- autoplot(prcomp(t_BP_norm_counts_LRT[-rev(seq_len(ncol(t_BP_norm_counts_LRT)))[1:2]]), data = t_BP_norm_counts_LRT) +
  geom_point(size=5,aes(shape=Treatment,color=Time_point)) +
  theme_bw(base_size=18) + 
  scale_color_manual(values=c("24h"="steelblue2","6h"="salmon","0h"="gray33")) +
  ggtitle("B. producta PCA") +
  theme(text = element_text(size=16)) +
  theme(aspect.ratio = 2/3) +
  stat_chull(aes(color = Time_point, fill = Time_point), 
             alpha = 0.1, geom = "polygon")
pdf('Figure_3A_BP_filtered.pdf')
Figure_3A_BP
dev.off()

## BS

BS_rld <- rlog(BS_dds, blind=FALSE)
BS_norm_counts <- as.data.frame(assay(BS_rld))

# visualize PCA using all counts
Supplementary_FIgure_3A_BS <- plotPCA(BS_rld, intgroup=c("BS_condition")) + theme_bw(base_size = 18) + geom_point(size=4) + theme(aspect.ratio = 1) + ggtitle('BS') + geom_text_repel(aes(label = name)) 

## make a custom PCAwith only significantly different genes using autoplot

# select only genes with a significant difference in any comparison
BS_LRT_dds <- DESeq(BS_dds, test="LRT", reduced=~1) 
BS_res <- as.data.frame(results(BS_LRT_dds))
BS_res$gene <- rownames(BS_res)
# make a vector with names of those significant genes
BS_res_sign <- filter(BS_res, padj < .05 & abs(log2FoldChange) > 1) 
BS_res_sign <- BS_res_sign$gene 
BS_norm_counts_LRT <- BS_norm_counts[which(rownames(BS_norm_counts) %in% BS_res_sign), ] 
BS_LRT_prcomp <- prcomp(x=t(BS_norm_counts_LRT), center = T, scale=T) #run PCA on the selected table of counts 
t_BS_norm_counts_LRT <- as.data.frame(t(BS_norm_counts_LRT)) %>% 
  mutate(Time_point=str_extract(rownames(BS_LRT_prcomp$x), pattern="0h|6h|24h"))  %>% 
  mutate(Treatment=str_extract(rownames(BS_LRT_prcomp$x), pattern="aCD3|flag")) %>% 
  replace_na(list(Treatment="None")) 


Figure_3A_BS <- autoplot(prcomp(t_BS_norm_counts_LRT[-rev(seq_len(ncol(t_BS_norm_counts_LRT)))[1:2]]), data = t_BS_norm_counts_LRT) +
  geom_point(size=5,aes(shape=Treatment,color=Time_point)) +
  theme_bw(base_size=18) + 
  scale_color_manual(values=c("24h"="steelblue2","6h"="salmon","0h"="gray33")) +
  ggtitle("B. sartorii PCA") +
  theme(text = element_text(size=16)) +
  theme(aspect.ratio = 2/3) +
  stat_chull(aes(color = Time_point, fill = Time_point), 
             alpha = 0.1, geom = "polygon")

pdf('Figure_3A_BS_filtered.pdf')
Figure_3A_BS
dev.off()

## CB

CB_rld <- rlog(CB_dds, blind=FALSE)
CB_norm_counts <- as.data.frame(assay(CB_rld))

# visualize PCA using all counts
Supplementary_FIgure_3A_CB <- plotPCA(CB_rld, intgroup=c("CB_condition")) + theme_bw(base_size = 18) + geom_point(size=4) + theme(aspect.ratio = 1) + ggtitle('CB') + geom_text_repel(aes(label = name)) 

## make a custom PCAwith only significantly different genes using autoplot

# select only genes with a significant difference in any comparison
CB_LRT_dds <- DESeq(CB_dds, test="LRT", reduced=~1) 
CB_res <- as.data.frame(results(CB_LRT_dds))
CB_res$gene <- rownames(CB_res)
# make a vector with names of those significant genes
CB_res_sign <- filter(CB_res, padj < .05 & abs(log2FoldChange) > 1) 
CB_res_sign <- CB_res_sign$gene 
CB_norm_counts_LRT <- CB_norm_counts[which(rownames(CB_norm_counts) %in% CB_res_sign), ] 
CB_LRT_prcomp <- prcomp(x=t(CB_norm_counts_LRT), center = T, scale=T) #run PCA on the selected table of counts 
t_CB_norm_counts_LRT <- as.data.frame(t(CB_norm_counts_LRT)) %>% 
  mutate(Time_point=str_extract(rownames(CB_LRT_prcomp$x), pattern="0h|6h|24h"))  %>% 
  mutate(Treatment=str_extract(rownames(CB_LRT_prcomp$x), pattern="aCD3|flag")) %>% 
  replace_na(list(Treatment="None")) 


Figure_3A_CB <- autoplot(prcomp(t_CB_norm_counts_LRT[-rev(seq_len(ncol(t_CB_norm_counts_LRT)))[1:2]]), data = t_CB_norm_counts_LRT) +
  geom_point(size=5,aes(shape=Treatment,color=Time_point)) +
  theme_bw(base_size=18) + 
  scale_color_manual(values=c("24h"="steelblue2","6h"="salmon","0h"="gray33")) +
  ggtitle("C. bolteae PCA") +
  theme(text = element_text(size=16)) +
  theme(aspect.ratio = 2/3) +
  stat_chull(aes(color = Time_point, fill = Time_point), 
             alpha = 0.1, geom = "polygon")


pdf('Figure_3A_CB_filtered.pdf')
Figure_3A_CB
dev.off()

## PD
PD_rld <- rlog(PD_dds, blind=FALSE)
PD_norm_counts <- as.data.frame(assay(PD_rld))

# visualize PCA using all counts
Supplementary_FIgure_3A_PD <- plotPCA(PD_rld, intgroup=c("PD_condition")) + theme_bw(base_size = 18) + geom_point(size=4) + theme(aspect.ratio = 1) + ggtitle('PD') + geom_text_repel(aes(label = name)) 

## make a custom PCAwith only significantly different genes using autoplot

# select only genes with a significant difference in any comparison
PD_LRT_dds <- DESeq(PD_dds, test="LRT", reduced=~1) 
PD_res <- as.data.frame(results(PD_LRT_dds))
PD_res$gene <- rownames(PD_res)
# make a vector with names of those significant genes
PD_res_sign <- filter(PD_res, padj < .05 & abs(log2FoldChange) > 1) 
PD_res_sign <- PD_res_sign$gene 
PD_norm_counts_LRT <- PD_norm_counts[which(rownames(PD_norm_counts) %in% PD_res_sign), ] 
PD_LRT_prcomp <- prcomp(x=t(PD_norm_counts_LRT), center = T, scale=T) #run PCA on the selected table of counts 
t_PD_norm_counts_LRT <- as.data.frame(t(PD_norm_counts_LRT)) %>% 
  mutate(Time_point=str_extract(rownames(PD_LRT_prcomp$x), pattern="0h|6h|24h"))  %>% 
  mutate(Treatment=str_extract(rownames(PD_LRT_prcomp$x), pattern="aCD3|flag")) %>% 
  replace_na(list(Treatment="None")) 


Figure_3A_PD <- autoplot(prcomp(t_PD_norm_counts_LRT[-rev(seq_len(ncol(t_PD_norm_counts_LRT)))[1:2]]), data = t_PD_norm_counts_LRT) +
  geom_point(size=5,aes(shape=Treatment,color=Time_point)) +
  theme_bw(base_size=18) + 
  scale_color_manual(values=c("24h"="steelblue2","6h"="salmon","0h"="gray33")) +
  ggtitle("P. distasonis PCA") +
  theme(text = element_text(size=16)) +
  theme(aspect.ratio = 2/3) +
  stat_chull(aes(color = Time_point, fill = Time_point), 
             alpha = 0.1, geom = "polygon")

pdf('Figure_3A_PD_filtered.pdf')
Figure_3A_PD
dev.off()


##### generate df containing DES for all BP comparisons ####
BP_flag6h_vs_0h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","6h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_flag6h_vs_0h") 
BP_aCD36h_vs_0h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","6h_aCD3","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_aCD36h_vs_0h") 
BP_aCD324h_vs_0h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","24h_aCD3", "0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_aCD324h_vs_0h") 
BP_flag24h_vs_0h  <- as.data.frame(results(BP_dds, contrast=c("BP_condition","24h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_flag24h_vs_0h") 
BP_aCD324h_vs_aCD36h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","24h_aCD3", "6h_aCD3"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_aCD324hh_vs_aCD36h") 
BP_flag24h_vs_flag6h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","24h_flag", "6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_flag24hh_vs_flag6h") 
BP_aCD36h_vs_flag6h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","6h_aCD3","6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_aCD36h_vs_flag6h")
BP_aCD324h_vs_flag24h <- as.data.frame(results(BP_dds, contrast=c("BP_condition","24h_aCD3", "24h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BP_aCD324h_vs_flag24h") 

#### make a unique df with all the BP comparisons
BP_complete_df <- rbind(BP_flag6h_vs_0h, BP_aCD36h_vs_0h, BP_aCD324h_vs_0h, BP_flag24h_vs_0h, BP_aCD36h_vs_flag6h, BP_aCD324h_vs_flag24h) %>% 
  mutate(Time_point = str_extract(Comparison, pattern ="6h|24h")) %>% 
  mutate(Treatment = case_when(
    Comparison %in% c("BP_flag6h_vs_0h", "BP_flag24h_vs_0h") ~ "flag",
    Comparison %in% c("BP_aCD36h_vs_0h", "BP_aCD324h_vs_0h") ~ "aCD3",
    Comparison %in% c("BP_aCD36h_vs_flag6h" ,"BP_aCD324h_vs_flag24h") ~ "aCD3vsflag"))  %>% 
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|33035.10.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))


##### generate df containing DES for all BS comparisons ####
BS_flag6h_vs_0h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","6h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_flag6h_vs_0h") 
BS_aCD36h_vs_0h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","6h_aCD3","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_aCD36h_vs_0h") 
BS_aCD324h_vs_0h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","24h_aCD3", "0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_aCD324h_vs_0h") 
BS_flag24h_vs_0h  <- as.data.frame(results(BS_dds, contrast=c("BS_condition","24h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_flag24h_vs_0h") 
BS_aCD324h_vs_aCD36h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","24h_aCD3", "6h_aCD3"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_aCD324hh_vs_aCD36h") 
BS_flag24h_vs_flag6h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","24h_flag", "6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_flag24hh_vs_flag6h") 
BS_aCD36h_vs_flag6h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","6h_aCD3","6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_aCD36h_vs_flag6h")
BS_aCD324h_vs_flag24h <- as.data.frame(results(BS_dds, contrast=c("BS_condition","24h_aCD3", "24h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="BS_aCD324h_vs_flag24h") 

#### make a unique df with all the BS comparisons
BS_complete_df <- rbind(BS_flag6h_vs_0h, BS_aCD36h_vs_0h, BS_aCD324h_vs_0h, BS_flag24h_vs_0h, BS_aCD36h_vs_flag6h, BS_aCD324h_vs_flag24h) %>% 
  mutate(Time_point = str_extract(Comparison, pattern ="6h|24h")) %>% 
  mutate(Treatment = case_when(
    Comparison %in% c("BS_flag6h_vs_0h", "BS_flag24h_vs_0h") ~ "flag",
    Comparison %in% c("BS_aCD36h_vs_0h", "BS_aCD324h_vs_0h") ~ "aCD3",
    Comparison %in% c("BS_aCD36h_vs_flag6h" ,"BS_aCD324h_vs_flag24h") ~ "aCD3vsflag"))  %>% 
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|671267.8.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))


##### generate df containing DES for all CB comparisons ####
CB_flag6h_vs_0h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","6h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_flag6h_vs_0h") 
CB_aCD36h_vs_0h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","6h_aCD3","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_aCD36h_vs_0h") 
CB_aCD324h_vs_0h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","24h_aCD3", "0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_aCD324h_vs_0h") 
CB_flag24h_vs_0h  <- as.data.frame(results(CB_dds, contrast=c("CB_condition","24h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_flag24h_vs_0h") 
CB_aCD324h_vs_aCD36h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","24h_aCD3", "6h_aCD3"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_aCD324hh_vs_aCD36h") 
CB_flag24h_vs_flag6h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","24h_flag", "6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_flag24hh_vs_flag6h") 
CB_aCD36h_vs_flag6h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","6h_aCD3","6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_aCD36h_vs_flag6h")
CB_aCD324h_vs_flag24h <- as.data.frame(results(CB_dds, contrast=c("CB_condition","24h_aCD3", "24h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="CB_aCD324h_vs_flag24h") 


#### make a unique df with all the CB comparisons
CB_complete_df <- rbind(CB_flag6h_vs_0h, CB_aCD36h_vs_0h, CB_aCD324h_vs_0h, CB_flag24h_vs_0h, CB_aCD36h_vs_flag6h, CB_aCD324h_vs_flag24h) %>% 
  mutate(Time_point = str_extract(Comparison, pattern ="6h|24h")) %>% 
  mutate(Treatment = case_when(
    Comparison %in% c("CB_flag6h_vs_0h", "CB_flag24h_vs_0h") ~ "flag",
    Comparison %in% c("CB_aCD36h_vs_0h", "CB_aCD324h_vs_0h") ~ "aCD3",
    Comparison %in% c("CB_aCD36h_vs_flag6h" ,"CB_aCD324h_vs_flag24h") ~ "aCD3vsflag"))  %>% 
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|208479.6.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))


##### generate df containing DES for all PD comparisons ####
PD_flag6h_vs_0h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","6h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_flag6h_vs_0h") 
PD_aCD36h_vs_0h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","6h_aCD3","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_aCD36h_vs_0h") 
PD_aCD324h_vs_0h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","24h_aCD3", "0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_aCD324h_vs_0h") 
PD_flag24h_vs_0h  <- as.data.frame(results(PD_dds, contrast=c("PD_condition","24h_flag","0h"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_flag24h_vs_0h") 
PD_aCD324h_vs_aCD36h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","24h_aCD3", "6h_aCD3"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_aCD324hh_vs_aCD36h") 
PD_flag24h_vs_flag6h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","24h_flag", "6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_flag24hh_vs_flag6h") 
PD_aCD36h_vs_flag6h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","6h_aCD3","6h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_aCD36h_vs_flag6h")
PD_aCD324h_vs_flag24h <- as.data.frame(results(PD_dds, contrast=c("PD_condition","24h_aCD3", "24h_flag"))) %>% rownames_to_column('Gene') %>% mutate(Comparison="PD_aCD324h_vs_flag24h") 

#### make a unique df with all the PD comparisons
PD_complete_df <- rbind(PD_flag6h_vs_0h, PD_aCD36h_vs_0h, PD_aCD324h_vs_0h, PD_flag24h_vs_0h, PD_aCD36h_vs_flag6h, PD_aCD324h_vs_flag24h) %>% 
  mutate(Time_point = str_extract(Comparison, pattern ="6h|24h")) %>% 
  mutate(Treatment = case_when(
    Comparison %in% c("PD_flag6h_vs_0h", "PD_flag24h_vs_0h") ~ "flag",
    Comparison %in% c("PD_aCD36h_vs_0h", "PD_aCD324h_vs_0h") ~ "aCD3",
    Comparison %in% c("PD_aCD36h_vs_flag6h" ,"PD_aCD324h_vs_flag24h") ~ "aCD3vsflag"))  %>% 
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|823.92.peg.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))



###########################################
############ CUMULATIVE PLOTS #############
###########################################

##### volcano plots flagellin Figure 1E ####

BP_f6 <- BP_flag6h_vs_0h %>% mutate(Organism='BP')
BS_f6 <- BS_flag6h_vs_0h %>% mutate(Organism='BS')
CB_f6 <- CB_flag6h_vs_0h %>% mutate(Organism='CB')
PD_f6 <- PD_flag6h_vs_0h %>% mutate(Organism='PD')

Flag_6h_all <- bind_rows(BP_f6, BS_f6, CB_f6, PD_f6)
  
flag_counts <- Flag_6h_all %>%
  dplyr::group_by(Organism) %>%
  dplyr::summarize(n.top=sum(padj<0.05 & log2FoldChange>1,na.rm=TRUE),
            n.bottom=sum(padj<0.05 & log2FoldChange< (-1),na.rm=TRUE)) %>%
  dplyr::ungroup()


Figure_1E <- Flag_6h_all %>%
  mutate(Significant = case_when(
    padj > 0.05 | (abs(log2FoldChange) < 1) ~ 'Not Sign',
    padj < 0.05 & log2FoldChange > 1 ~ 'UP',
    padj < 0.05 & log2FoldChange <(-1) ~ 'DOWN'
  )) %>% 
  arrange(Significant!="Not Sign") %>%
  ggplot() + 
  geom_point(aes(x=log2FoldChange, y=-log10(pvalue), color=Significant), size=3) + theme_bw(base_size=16) +
  scale_color_manual(values=c("UP"="firebrick2","Not Sign"="darkgrey", 'DOWN'='dodgerblue3')) +   
  geom_vline(xintercept = -1, colour="#990000", linetype="dashed") +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") +
  facet_wrap(~factor(Organism, levels=c('CB', 'BP', 'BS', 'PD')), nrow=2) +
  geom_text(data=flag_counts,aes(x=4, y=22,  label=n.top, size=6)) + 
  geom_text(data=flag_counts,aes(x=-3, y=22,  label=n.bottom,  size=6)) +
  theme(strip.background = element_blank(), strip.text.y = element_blank()) 

pdf('Figure_1E_filtered.pdf')
Figure_1E
dev.off()

##### volcano plots aCD3 Figure 2C ####

BP_aCD36 <- BP_aCD36h_vs_0h %>% mutate(Organism='BP')
BS_aCD36 <- BS_aCD36h_vs_0h %>% mutate(Organism='BS')
CB_aCD36 <- CB_aCD36h_vs_0h %>% mutate(Organism='CB')
PD_aCD36 <- PD_aCD36h_vs_0h %>% mutate(Organism='PD')

aCD3_6h_all <- bind_rows(BP_aCD36, BS_aCD36, CB_aCD36, PD_aCD36)

aCD3_counts <- aCD3_6h_all %>%
  dplyr::group_by(Organism) %>%
  dplyr::summarize(n.top=sum(padj<0.05 & log2FoldChange>1,na.rm=TRUE),
                   n.bottom=sum(padj<0.05 & log2FoldChange< (-1),na.rm=TRUE)) %>%
  dplyr::ungroup()


Figure_2C <- aCD3_6h_all %>%
  drop_na(padj) %>%
  mutate(Significant = case_when(
    padj > 0.05 | (abs(log2FoldChange) < 1) ~ 'Not Sign',
    padj < 0.05 & log2FoldChange > 1 ~ 'UP',
    padj < 0.05 & log2FoldChange <(-1) ~ 'DOWN'
  )) %>% 
  arrange(Significant!="Not Sign") %>%
  ggplot() + 
  geom_point(aes(x=log2FoldChange, y=-log10(pvalue), color=Significant), size=3) + theme_bw(base_size=16) +
  scale_color_manual(values=c("UP"="firebrick2","Not Sign"="darkgrey", 'DOWN'='dodgerblue3')) +   
  geom_vline(xintercept = -1, colour="#990000", linetype="dashed") +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") +
  facet_wrap(~factor(Organism, levels=c('CB', 'BP', 'BS', 'PD')), nrow=2) +
  geom_text(data=aCD3_counts,aes(x=4, y=22,  label=n.top, size=6)) + 
  geom_text(data=aCD3_counts,aes(x=-3, y=22,  label=n.bottom,  size=6)) +
  theme(strip.background = element_blank(), strip.text.y = element_blank()) 

pdf('Figure_2C_filtered.pdf')
Figure_2C
dev.off()



##### comparison dot-plot aCD3 flagellin Figure 3B #####

BP_all_6h <- BP_aCD36h_vs_0h %>% full_join(BP_flag6h_vs_0h, by='Gene') %>% full_join(BP_aCD36h_vs_flag6h, by='Gene') %>%
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|33035.10.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))

BS_all_6h <- BS_aCD36h_vs_0h %>% full_join(BS_flag6h_vs_0h, by='Gene') %>% full_join(BS_aCD36h_vs_flag6h, by='Gene') %>%
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|33035.10.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))

CB_all_6h <- CB_aCD36h_vs_0h %>% full_join(CB_flag6h_vs_0h, by='Gene') %>% full_join(CB_aCD36h_vs_flag6h, by='Gene') %>%
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|33035.10.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))

PD_all_6h <- PD_aCD36h_vs_0h %>% full_join(PD_flag6h_vs_0h, by='Gene') %>% full_join(PD_aCD36h_vs_flag6h, by='Gene') %>%
  mutate(Gene_name = gsub("(.*);Name=", "", Gene)) %>% mutate(Gene_name = gsub("(\\(E.*)", "", Gene_name)) %>%
  mutate(Gene_number = gsub("(;.*)" , "", Gene)) %>% mutate(Gene_number = gsub("ID=fig\\|33035.10.", "", Gene_number)) %>% mutate(Gene_number = gsub("\\|", "", Gene_number))


BP_aCD3_vs_flag_6h_forgg <- BP_all_6h %>% mutate(Rel_Sign = case_when(
  padj < 0.05 ~ "true_dif",
  padj.x < 0.05 & padj.y < 0.05 ~ "both",
  padj.x < 0.05 & (padj.y > 0.05 | padj.y==NA) ~ "aCD3",
  (padj.x > 0.05 | padj.x==NA) & padj.y < 0.05 ~ "flag",
  (padj.x > 0.05 | padj.x==NA) & (padj.y > 0.05 | padj.y==NA) ~ "none")) %>%
  mutate(FC_Sign = case_when(
    log2FoldChange.x > 0 & log2FoldChange.y < 0 ~ "disc, aCD>0", 
    log2FoldChange.x < 0 & log2FoldChange.y > 0 ~ "disc, flag>0",
    log2FoldChange.x > 0 & log2FoldChange.y > 0  ~ "both >0",
    log2FoldChange.x < 0 & log2FoldChange.y < 0 ~ "both <0"))

BS_aCD3_vs_flag_6h_forgg <- BS_all_6h %>% mutate(Rel_Sign = case_when(
  padj < 0.05 ~ "true_dif",
  padj.x < 0.05 & padj.y < 0.05 ~ "both",
  padj.x < 0.05 & (padj.y > 0.05 | padj.y==NA) ~ "aCD3",
  (padj.x > 0.05 | padj.x==NA) & padj.y < 0.05 ~ "flag",
  (padj.x > 0.05 | padj.x==NA) & (padj.y > 0.05 | padj.y==NA) ~ "none")) %>%
  mutate(FC_Sign = case_when(
    log2FoldChange.x > 0 & log2FoldChange.y < 0 ~ "disc, aCD>0", 
    log2FoldChange.x < 0 & log2FoldChange.y > 0 ~ "disc, flag>0",
    log2FoldChange.x > 0 & log2FoldChange.y > 0  ~ "both >0",
    log2FoldChange.x < 0 & log2FoldChange.y < 0 ~ "both <0"))

CB_aCD3_vs_flag_6h_forgg <- CB_all_6h %>% mutate(Rel_Sign = case_when(
  padj < 0.05 ~ "true_dif",
  padj.x < 0.05 & padj.y < 0.05 ~ "both",
  padj.x < 0.05 & (padj.y > 0.05 | padj.y==NA) ~ "aCD3",
  (padj.x > 0.05 | padj.x==NA) & padj.y < 0.05 ~ "flag",
  (padj.x > 0.05 | padj.x==NA) & (padj.y > 0.05 | padj.y==NA) ~ "none")) %>%
  mutate(FC_Sign = case_when(
    log2FoldChange.x > 0 & log2FoldChange.y < 0 ~ "disc, aCD>0", 
    log2FoldChange.x < 0 & log2FoldChange.y > 0 ~ "disc, flag>0",
    log2FoldChange.x > 0 & log2FoldChange.y > 0  ~ "both >0",
    log2FoldChange.x < 0 & log2FoldChange.y < 0 ~ "both <0"))

PD_aCD3_vs_flag_6h_forgg <- PD_all_6h %>% mutate(Rel_Sign = case_when(
  padj < 0.05 ~ "true_dif",
  padj.x < 0.05 & padj.y < 0.05 ~ "both",
  padj.x < 0.05 & (padj.y > 0.05 | padj.y==NA) ~ "aCD3",
  (padj.x > 0.05 | padj.x==NA) & padj.y < 0.05 ~ "flag",
  (padj.x > 0.05 | padj.x==NA) & (padj.y > 0.05 | padj.y==NA) ~ "none")) %>%
  mutate(FC_Sign = case_when(
    log2FoldChange.x > 0 & log2FoldChange.y < 0 ~ "disc, aCD>0", 
    log2FoldChange.x < 0 & log2FoldChange.y > 0 ~ "disc, flag>0",
    log2FoldChange.x > 0 & log2FoldChange.y > 0  ~ "both >0",
    log2FoldChange.x < 0 & log2FoldChange.y < 0 ~ "both <0"))

all_aCD3_vs_flag_6h_forgg <- bind_rows(BP_aCD3_vs_flag_6h_forgg, BS_aCD3_vs_flag_6h_forgg, CB_aCD3_vs_flag_6h_forgg, PD_aCD3_vs_flag_6h_forgg) %>%
  mutate(Rel_Sign = case_when(
    # padj < 0.05 ~ "true_dif",
    padj.x < 0.05 & padj.y < 0.05 ~ "both",
    padj.x < 0.05 & (padj.y > 0.05 | padj.y==NA) ~ "aCD3",
    (padj.x > 0.05 | padj.x==NA) & padj.y < 0.05 ~ "flag",
    (padj.x > 0.05 | padj.x==NA) & (padj.y > 0.05 | padj.y==NA) ~ "none")) %>%
  mutate(FC_Sign = case_when(
    log2FoldChange.x > 0 & log2FoldChange.y < 0 ~ "disc, aCD>0", 
    log2FoldChange.x < 0 & log2FoldChange.y > 0 ~ "disc, flag>0",
    log2FoldChange.x > 0 & log2FoldChange.y > 0  ~ "both >0",
    log2FoldChange.x < 0 & log2FoldChange.y < 0 ~ "both <0")) %>%
  mutate(Organism = str_extract(pattern='BS|BP|CB|PD', Comparison)) %>%
  arrange(Rel_Sign!="none",Rel_Sign!="aCD3",Rel_Sign!="flag") %>%
  mutate(alpha=ifelse(Rel_Sign=="none",0.6,1))


pick <- function(condition){
  function(d) d %>% filter_(condition)
}

Figure_3B <- 
  #filter(Rel_Sign != "none") %>%
  ggplot(all_aCD3_vs_flag_6h_forgg, aes(x=log2FoldChange.x, y=log2FoldChange.y, colour=Rel_Sign,alpha=alpha)) +
  stat_cor(inherit.aes = FALSE, mapping = aes(x=log2FoldChange.x, y=log2FoldChange.y), method="pearson", size=6) + #this is necessary as otherwise with inherited aes the value is calculated separately for the color groups!
  # geom_text_repel(data=filter(all_aCD3_vs_flag_24h_forgg, Gene %in% list_glut), segment.size  = 0.4, 
  #                 aes(label=Gene_name)) +
  #geom_hline(yintercept = 0, colour="#990000", linetype="dashed") +
  #geom_vline(xintercept = 0, coloFigure 3Fur="#990000", linetype="dashed") +
  geom_abline(slope=1, colour="#990000", linetype="dashed") +
  geom_point(size=3) +
  scale_alpha_identity() +
  scale_colour_manual(values=c( "none"="darkgrey", "both"= "plum3","aCD3"="red", "flag"="#00B2FF")) +
  theme_bw(base_size=18) + 
  theme(aspect.ratio = 1) +
  #xlim(-7.5, 7.5) +
  ggtitle("all 6h") +
  xlab("log2FC in aCD3") +
  ylab("log2FC in flagellin") +
  theme(legend.text = element_text(size = 20),
        strip.background = element_blank()) +
  facet_wrap(~ Organism, scales = 'free', nrow=1) 


pdf('Figure_3B_filtered.pdf')
Figure_3B
dev.off()


##### volcanoplots Figure 3C and Supplementary Figure 3B  #####

all <- bind_rows(BP_complete_df, BS_complete_df, CB_complete_df, PD_complete_df) %>%
  filter(!grepl('aCD36h_vs_flag6h|aCD324h_vs_flag24h', Comparison)) %>%
  mutate(Organism = str_extract(pattern='BP|BS|CB|PD', Comparison)) %>%
  #mutate(padj = coalesce(padj, 1)) %>%
  mutate(Significant = case_when(
    padj>0.05 | abs(log2FoldChange)<1 ~ 'not_sign',
    padj<0.05 & abs(log2FoldChange)>1 ~ 'sign'))

all_6h <-  all %>%
  filter(Time_point=='6h') 

 chaperone_6h <- all_6h %>%
   ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
   geom_point(aes(color = Significant), size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Hsp|Chaperone|chaperone|ClpB"), all_6h$Gene_name)) %>% filter(Significant=='not_sign'), fill="goldenrod1", color="black", pch=21, size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Hsp|Chaperone|chaperone|ClpB"), all_6h$Gene_name)) %>% filter(Significant=='sign'), fill="cornflowerblue", color="black", pch=21, size=3) +
   scale_color_manual(values = c("grey85", "grey70")) +
   theme_bw(base_size = 12) + theme(legend.position = "bottom") +
   ggtitle("Chaperones") +
   facet_wrap( Organism~Treatment, ncol=2) +
   theme(strip.background = element_blank(),
         strip.text.y = element_blank(),
         strip.text.x = element_blank(), 
         aspect.ratio = 1) 

 
 gg_PUL_6h <- all_6h %>%
   ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
   geom_point(aes(color = Significant), size=3) +
   geom_point(data = all_6h %>% filter(grepl(("(PUL)"), all_6h$Gene_name)) %>% filter(Significant=='not_sign'), fill="goldenrod1", color="black", pch=21, size=3) +
   geom_point(data = all_6h %>% filter(grepl(("(PUL)"), all_6h$Gene_name)) %>% filter(Significant=='sign'), fill="cornflowerblue", color="black", pch=21, size=3) +
   scale_color_manual(values = c("grey85", "grey70")) +
   theme_bw(base_size = 12) + theme(legend.position = "bottom") +
   ggtitle("Polysaccharide Utilization Loci") +
   facet_wrap( Organism~Treatment, ncol=2) +
   theme(strip.background = element_blank(),
         strip.text.y = element_blank(),
         strip.text.x = element_blank(), 
         aspect.ratio = 1) 


 gg_ferric_iron_transporters_6h <- all_6h %>%
   ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
   geom_point(aes(color = Significant), size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Ferric iron | ferric iron"), all_6h$Gene_name)) %>% filter(Significant=='not_sign'), fill="goldenrod1", color="black", pch=21, size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Ferric iron | ferric iron"), all_6h$Gene_name)) %>% filter(Significant=='sign'), fill="cornflowerblue", color="black", pch=21, size=3) +
   scale_color_manual(values = c( "grey85", "grey70")) +
   theme_bw(base_size = 12) + theme(legend.position = "bottom") +
   ggtitle("Ferric iron uptake (ABC transporters and associated two-component systems)") +
   facet_wrap( Treatment~Organism, nrow=2) +
   theme(strip.background = element_blank(),
         strip.text.y = element_blank(),
         strip.text.x = element_blank(),
         aspect.ratio = 1)

 
 
 gg_catalases_6h <- all_6h %>%
   ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
   geom_point(aes(color = Significant), size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Catalase|catalase"), all_6h$Gene_name)) %>% filter(Significant=='not_sign'), fill="goldenrod1", color="black", pch=21, size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Catalase|catalase"), all_6h$Gene_name)) %>% filter(Significant=='sign'), fill="cornflowerblue", color="black", pch=21, size=3) +
   scale_color_manual(values = c( "grey85", "grey70")) +
   theme_bw(base_size = 12) + theme(legend.position = "bottom") +
   ggtitle("Catalases") +
   facet_wrap( Organism~Treatment, ncol=2) +
   theme(strip.background = element_blank(),
         strip.text.y = element_blank(),
         strip.text.x = element_blank(), 
         aspect.ratio = 1) 
 
 
 
 gg_thioredoxin_6h <- all_6h %>%
   ggplot(aes(x = log2FoldChange, y = -log10(pvalue))) +
   geom_point(aes(color = Significant), size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Thioredoxin|thioredoxin"), all_6h$Gene_name)) %>% filter(Significant=='not_sign'), fill="goldenrod1", color="black", pch=21, size=3) +
   geom_point(data = all_6h %>% filter(grepl(("Thioredoxin|thioredoxin"), all_6h$Gene_name)) %>% filter(Significant=='sign'), fill="cornflowerblue", color="black", pch=21, size=3) +
   scale_color_manual(values = c( "grey85", "grey70")) +
   theme_bw(base_size = 12) + theme(legend.position = "bottom") +
   ggtitle("Thioredoxin") +
   facet_wrap( Treatment~Organism, nrow=2) +
   theme(strip.background = element_blank(),
         strip.text.y = element_blank(),
         strip.text.x = element_blank(),
         aspect.ratio = 1) 
 
 
 
##### housekeeping genes selection for qPCR ####
 
 SB093_BP_HK_genes <- BP_norm_counts %>% mutate(variance=rowVars(BP_norm_counts)) %>% 
   mutate(meancounts=rowMeans(BP_norm_counts)) %>%
   rownames_to_column('Gene') %>% 
   filter(variance < quantile(variance, .02)) %>%# & meancounts > quantile(meancounts, .50)) %>%
   select(Gene, meancounts, variance) %>%
   mutate(Exp='SB093') %>% mutate(Organism='BP') %>% arrange(variance) 
 
 SB093_BS_HK_genes <- BS_norm_counts %>% mutate(variance=rowVars(BS_norm_counts)) %>% 
   mutate(meancounts=rowMeans(BS_norm_counts)) %>%
   rownames_to_column('Gene') %>% 
   filter(variance < quantile(variance, .02)) %>%# & meancounts > quantile(meancounts, .50)) %>%
   select(Gene, meancounts, variance) %>%
   mutate(Exp='SB093') %>% mutate(Organism='BS') %>% arrange(variance) 
 
 SB093_CB_HK_genes <- CB_norm_counts %>% mutate(variance=rowVars(CB_norm_counts)) %>% 
   mutate(meancounts=rowMeans(CB_norm_counts)) %>%
   rownames_to_column('Gene') %>% 
   filter(variance < quantile(variance, .02))%>%# & meancounts > quantile(meancounts, .50)) %>%
   select(Gene, meancounts, variance) %>%
   mutate(Exp='SB093') %>% mutate(Organism='CB') %>% arrange(variance) 
 
 SB093_PD_HK_genes <- PD_norm_counts %>% mutate(variance=rowVars(PD_norm_counts)) %>% 
   mutate(meancounts=rowMeans(PD_norm_counts)) %>%
   rownames_to_column('Gene') %>% 
   filter(variance < quantile(variance, .02)) %>%# & meancounts > quantile(meancounts, .50)) %>%
   select(Gene, meancounts, variance) %>%
   mutate(Exp='SB093') %>% mutate(Organism='PD') %>% arrange(variance) 
 
 bind_rows(SB093_BP_HK_genes, SB093_BS_HK_genes, SB093_CB_HK_genes, SB093_PD_HK_genes) %>%
   write.table('SB093_HK_genes.txt',quote=F, row.names=F, sep='\t')
 

##### get list of top 100 highly expressed genes in FASTA header format for Supplementary Figure 1 #####
 
 BP_counts_baseline <- BP_norm_counts %>% select(1,2,3) %>% mutate(mean_counts = rowMeans(BP_norm_counts)) %>% select(4) %>% mutate(Organism = 'BP') %>% rownames_to_column('Gene') %>% mutate(Gene=gsub('ID=', '>', Gene)) %>% mutate(Gene=gsub("(;.*)" , "", Gene)) %>% arrange( desc(mean_counts)) %>% head(100) %>% select(Gene) #%>% write.table('SB093_BP_counts_baseline_top100.txt', quote=FALSE, row.names=F, col.names=F)
 BS_counts_baseline <- BS_norm_counts %>% select(1,2,3) %>% mutate(mean_counts = rowMeans(BS_norm_counts)) %>% select(4) %>% mutate(Organism = 'BS') %>% rownames_to_column('Gene') %>% mutate(Gene=gsub('ID=', '>', Gene)) %>% mutate(Gene=gsub("(;.*)" , "", Gene)) %>% arrange( desc(mean_counts)) %>% head(100) %>% select(Gene) #%>% write.table('SB093_BS_counts_baseline_top100.txt', quote=FALSE, row.names=F, col.names=F)
 CB_counts_baseline <- CB_norm_counts %>% select(1,2,3) %>% mutate(mean_counts = rowMeans(CB_norm_counts)) %>% select(4) %>% mutate(Organism = 'CB') %>% rownames_to_column('Gene') %>% mutate(Gene=gsub('ID=', '>', Gene)) %>% mutate(Gene=gsub("(;.*)" , "", Gene)) %>% arrange( desc(mean_counts)) %>% head(100) %>% select(Gene) #%>% write.table('SB093_CB_counts_baseline_top100.txt', quote=FALSE, row.names=F, col.names=F)
 PD_counts_baseline <- PD_norm_counts %>% select(1,2,3) %>% mutate(mean_counts = rowMeans(PD_norm_counts)) %>% select(4) %>% mutate(Organism = 'PD') %>% rownames_to_column('Gene') %>% mutate(Gene=gsub('ID=', '>', Gene)) %>% mutate(Gene=gsub("(;.*)" , "", Gene)) %>% arrange( desc(mean_counts)) %>% head(100) %>% select(Gene) #%>% write.table('SB093_PD_counts_baseline_top100.txt', quote=FALSE, row.names=F, col.names=F)
 


 